File: [0322][Medium][Coin_Change]/Coin_Change.cpp
// leetcode time     cost : 52 ms
// leetcode memory   cost : 16.5 MB 
// Time  Complexity: O(S*n), S is the needed money amount
// Space Complexity: O(S)
// https://leetcode.com/problems/coin-change/discuss/77373/6-7-lines-2-ways
// slolution 1, coins outer loop, average 92ms
#include <vector>
#include <iostream>
using namespace std;
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        //new a vector need[amount+1] and init all its elements to the max value amount+1.
        // need[amount+1] is the smallest Num of coins when the need money is amount.
        vector<int> need(amount+1, amount+1);
        need[0] = 0;
        for (int coin : coins)
            for (int amount_x=coin; amount_x<=amount; amount_x++){
                need[amount_x] = min(need[amount_x], need[amount_x-coin] + 1);
                cout << "amount_x: "<< amount_x<<",need[amount_x] value is " << (need[amount_x])<< '\n';
            }
        return need.back() > amount ? -1 : need.back();
    }
};

int main(){
    vector<int> coins = {1, 2, 5};   // expect is 3, 11 = 5 + 5 + 1
    int amount = 11;
    Solution *Solution_obj = new Solution();
    int result = Solution_obj->coinChange(coins, amount);
	  cout << "In cpp code,result value is " << (result)<< '\n';
    return 0;
}

File: [0322][Medium][Coin_Change]/Coin_Change.py
# leetcode time     cost : 1460  ms
# leetcode memory   cost : 13.7 MB 
# Time  Complexity: O(n*k)
# Space Complexity: O(n)
# slolution 1, Dynamic programming - Bottom up,loop money amount with iteration
class Solution(object):
    # def coinChange(self, coins: List[int], amount: int) -> int:
    def coinChange(self, coins, amount):
        rs = [amount+1] * (amount+1)
        rs[0] = 0
        for i in range(1, amount+1):
            for c in coins:
                if i >= c:
                    rs[i] = min(rs[i], rs[i-c] + 1)

        if rs[amount] == amount+1:
            return -1
        return rs[amount]
# leetcode time     cost : 1012 ms
# leetcode memory   cost : 13.5 MB     
    def coinChange_2(self, coins: 'List[int]', amount: 'int') -> 'int':
            dp = [0] + [float('inf') for i in range(amount)]
            for i in range(1, amount+1):
                for coin in coins:
                    if i - coin >= 0:
                        dp[i] = min(dp[i], dp[i-coin] + 1)
            if dp[-1] == float('inf'):
                return -1
            return dp[-1]